test/runtests.jl

```diff
@@ -116,6 +118,53 @@ ch2 = chain(1:0, 1:2:5, 0.2:0.1:1.6)
 
 @test eltype(ch2) == typejoin(Int, Float64)
 @test collect(ch2) == [1:2:5; 0.2:0.1:1.6]
+@test length(ch2) == length(collect(ch2))
+if VERSION >= v"0.5.0-dev+3305"
+	@test iteratorsize(ch2) == HasLength()
+end
+
+ch3 = chain(1:10, 1:10, 1:10)
+@test length(ch3) == 30
+if VERSION >= v"0.5.0-dev+3305"
+	@test iteratorsize(ch3) == HasLength()
+end
+
+r = countfrom(1)
+ch4 = chain(1:10, countfrom(1))
+@test_throws MethodError length(ch4)
+if VERSION >= v"0.5.0-dev+3305"
+	@assert iteratorsize(r) == IsInfinite()
+	@test iteratorsize(ch4) == IsInfinite()
+end
+
+ch5 = chain()
+@test length(ch5) == 0
+if VERSION >= v"0.5.0-dev+3305"
+	@test iteratorsize(ch5) == HasLength()
+end
+
+c = chain(ch1, ch2, ch3)
+@test length(c) == length(ch1) + length(ch2) + length(ch3)
+#@test collect(c) == [ch1; ch2; ch3] # why doesn't work?
+@test collect(c) == [collect(ch1); collect(ch2); collect(ch3)]
+
+r = rand(2,2)
+c = chain(r, r)
+@test length(c) == 8
+@test collect(c) == [vec(r); vec(r)]
+if VERSION >= v"0.5.0-dev+3305"
+	@test iteratorsize(r) == HasShape()
+	@test iteratorsize(c) == HasLength()
+end
+
+if VERSION >= v"0.5.0-dev+3305"
+	r = distinct(1:10)
```

Comment:
Use `distinct(collect(1:10))`; it's possible someone might make a no-op `UnitRange` optimization for `distinct`.